pkg/controllers/node/node_controller.go

```diff
@@ -0,0 +1,283 @@
+
+	log "github.com/sirupsen/logrus"
+	"k8s.io/apimachinery/pkg/fields"
+	uruntime "k8s.io/apimachinery/pkg/util/runtime"
+	"k8s.io/client-go/kubernetes"
+	"k8s.io/client-go/pkg/api/v1"
+	corecache "k8s.io/client-go/tools/cache"
+
+	"fmt"
+)
+
+// This cache contains kubernetes node information keyed by etcd node-name.
+type cache struct {
+	sync.RWMutex
+	// The key for the cache's data is an etcd nodeName.
+	data map[string]string
+}
+
+// NodeController implements the Controller interface.  It is responsible for monitoring
+// kubernetes nodes and responding to delete events by removing them from the Calico datastore.
+// It keeps a cache of known calico nodes and their corresponding kubernetes nodes to
+// accomplish this.
+type NodeController struct {
+	indexer       corecache.Indexer
+	informer      corecache.Controller
+	nodeDataCache calicocache.ResourceCache
+	nodeCache *cache
+	calicoClient  *client.Client
+	k8sClientset  *kubernetes.Clientset
+}
+
+
+// NewNodeController Constructor for NodeController
+func NewNodeController(k8sClientset *kubernetes.Clientset, calicoClient *client.Client) controller.Controller {
+	cacheArgs := calicocache.ResourceCacheArgs{
+		ObjectType: reflect.TypeOf(""),
+		ListFunc:   func() (map[string]interface{}, error) {
+			// Get all nodes from the Calico datastore
+			calicoNodes, err := calicoClient.Nodes().List(api.NodeMetadata{})
+			if err != nil {
+				return nil, err
+			}
+
+			// Store the relevant information in our cache.
+			m := make(map[string]interface{})
+			for _, calicoNode := range calicoNodes.Items {
+				// find its kubernetes orchRef
+				k8sNodeName, err := getK8sNodeRef(calicoNode)
+				if err != nil {
+					m[k8sNodeName] = ""
+				}
+			}
+
+			log.Debugf("Found %d nodes in Calico datastore:", len(m))
+			return m, nil
+		},
+	}
+
+	nodeDataCache := calicocache.NewResourceCache(cacheArgs)
+	nodeCache := cache{data: make(map[string]string)}
+
+	// Create a Node watcher.
+	listWatcher := corecache.NewListWatchFromClient(k8sClientset.Core().RESTClient(), "nodes", "", fields.Everything())
+
+	// Bind the Calico cache to kubernetes cache with the help of an informer. This way we make sure that
+	// whenever the kubernetes cache is updated, changes get reflected in the Calico cache as well.
+	indexer, informer := corecache.NewIndexerInformer(listWatcher, &v1.Node{}, 0, corecache.ResourceEventHandlerFuncs{
+		DeleteFunc: func(obj interface{}) {
+			node, err := convert(obj)
+			if err != nil {
+				log.WithError(err).Errorf("Error while converting %#v to k8s node", node)
+			}
+			log.Debugf("Got DELETE event for node: %s", node)
+			nodeDataCache.Delete(node.ObjectMeta.Name)
+		},
+
+		AddFunc: func(obj interface{}) {
+			log.Debug("Add func.")
+			node, err := convert(obj)
+			if err != nil {
+				log.WithError(err).Errorf("Error while converting %#v to k8s node", node)
+			}
+			// Use an empty value here because the k8s node is created before the etcd node (so we can't find
```

Comment:
This comment isn't quite right - we use an empty value here because the only thing we care about it the kubernete's node's name, so there's no other relevant information we want to store in the cache besides the name.